server/handlers/validation.go

```diff
@@ -193,71 +202,79 @@ func loadTargetsFromStore(gun, role string, repo *tuf.Repo, store storage.MetaSt
-	}
+	case nil:
+		return metaUpdate, nil
 
-	currentJSON, err := store.GetCurrent(gun, data.CanonicalSnapshotRole)
-	if err != nil {
-		if _, ok := err.(storage.ErrNotFound); !ok {
-			return nil, validation.ErrValidation{Msg: err.Error()}
-		}
+	default:
+		return nil, validation.ErrValidation{Msg: err.Error()}
 	}
-	var sn *data.SignedSnapshot
-	if currentJSON != nil {
-		sn = new(data.SignedSnapshot)
-		err := json.Unmarshal(currentJSON, sn)
-		if err != nil {
-			return nil, validation.ErrValidation{Msg: err.Error()}
-		}
-		err = repo.SetSnapshot(sn)
-		if err != nil {
-			return nil, validation.ErrValidation{Msg: err.Error()}
-		}
-	} else {
-		// this will only occurr if no snapshot has ever been created for the repository
-		err := repo.InitSnapshot()
-		if err != nil {
-			return nil, validation.ErrBadSnapshot{Msg: err.Error()}
+}
+
+// generateTimestamp generates a new timestamp from the previous one in the store - this assumes all
+// the other roles have already been set on the repo, and will set the generated timestamp on the repo as well
+func generateTimestamp(gun string, repo *tuf.Repo, store storage.MetaStore) (*storage.MetaUpdate, error) {
+	var prev *data.SignedTimestamp
+	currentJSON, err := store.GetCurrent(gun, data.CanonicalTimestampRole)
+	if err == nil {
+		prev = new(data.SignedTimestamp)
+		if err = json.Unmarshal(currentJSON, prev); err != nil {
+			logrus.Error("Failed to unmarshal existing timestamp for GUN ", gun)
+			return nil, err
 		}
 	}
-	sgnd, err := repo.SignSnapshot(data.DefaultExpires(data.CanonicalSnapshotRole))
-	if err != nil {
-		return nil, validation.ErrBadSnapshot{Msg: err.Error()}
+
+	if _, ok := err.(storage.ErrNotFound); !ok && err != nil {
+		return nil, validation.ErrValidation{Msg: err.Error()}
 	}
-	sgndJSON, err := json.Marshal(sgnd)
+
+	metaUpdate, err := timestamp.NewTimestampUpdate(prev, repo)
 	if err != nil {
-		return nil, validation.ErrBadSnapshot{Msg: err.Error()}
+		_, noSigs := err.(signed.ErrInsufficientSignatures)
+		_, noKeys := err.(signed.ErrNoKeys)
+		// If we cannot sign the timestamp, then we don't have keys for the timestamp,
+		// and the client screwed up their root
+		if noSigs || noKeys {
+			return nil, validation.ErrBadRoot{
+				Msg: fmt.Sprintf("none of the following timestamp keys exist on the server: %s",
+					strings.Join(repo.Root.Signed.Roles[data.CanonicalTimestampRole].KeyIDs, ", ")),
+			}
+		}
+
+		return nil, validation.ErrValidation{Msg: err.Error()}
 	}
-	return &storage.MetaUpdate{
-		Role:    data.CanonicalSnapshotRole,
-		Version: repo.Snapshot.Signed.Version,
-		Data:    sgndJSON,
-	}, nil
+	return metaUpdate, nil
 }
 
+// validateSnapshot validates that the given snapshot update is valid.  It also sets the new snapshot
+// on the TUF repo, if it is valid
 func validateSnapshot(role string, oldSnap *data.SignedSnapshot, snapUpdate storage.MetaUpdate, roles map[string]storage.MetaUpdate, repo *tuf.Repo) error {
```

Comment:
it seems like we should rename this method to `validateAndLoadSnapshot` (or `loadAndValidateSnapshot` to be consistent with targets), since it is now setting snapshot on the repo